Fix 3xx responses from upstream

danlsgiga opened the pull request:
Close #33 

- Reply to the client with the correct response status code
- Use a better, more specific `defaultCacheKeyTemplate` to exact matches on resources that are redirected.
- The other changes are my VSCode editor removing white spaces
- Changes how the file backend store the files by using sha256 hashes based on the entry keyWithRespectVary() content. This way, we don't waste storage space by creating duplicated temporary files every time the caddy process is restarted.
- Implemented the Length() function for the file backend, since some browsers have grief when the Content-Length is set to 0 and abort the operation causing the http request to have an empty body
- Fix some tests

Commit:
fix(handler): handle 3xx responses from upstream

Commit:
fix(caddyfile): Define a better defaultKeyTemplate

Commit:
fix(endpoint): Fix tests

Commit:
fix(endpoint): Another test fix try

Commit:
fix(endpoint): Fix URI parsing

Commit:
fix(cdp-cache): Fix Content-Length for file backend

Commit:
fix(cdp-cache): Fix cache put ordering

- Filter out "stream closed" logs

fgilio commented:
Hi @danlsgiga!
After reading your issue  I wonder, does this PR make caching work for 30x responses or does it completely disable catching on those?

danlsgiga commented:
hi @fgilio, yeah... basically, 3xx responses are redirection only and what was happening is that I was getting 302's from the upstream but the cache was responding with a 200 and the http redirection never happened since the browser only reads the new `Location:` header if the response is 301 or 302.

danlsgiga commented:
Also, I'm sorry for the failing tests... I started using the memory backend but soon I started to get huge memory usages, probably memory leaking too, and moved to the file backend and made a few adjustments there too.

sillygod commented:
> hi @fgilio, yeah... basically, 3xx responses are redirection only and what was happening is that I was getting 302's from the upstream but the cache was responding with a 200 and the http redirection never happened since the browser only reads the new `Location:` header if the response is 301 or 302.

Hi @danlsgiga ,

Thanks for your work. I've fixed this bug in another pull request so this pr need to be updated.

sillygod commented on caddyfile.go:
```diff
@@ -43,7 +43,7 @@ var (
 	defaultCacheMaxMemorySize     = GB // default is 1 GB
 	defaultRedisConnectionSetting = "localhost:6379 0"
-	defaultCacheKeyTemplate       = "{http.request.method} {http.request.host}{http.request.uri.path}?{http.request.uri.query}"
+	defaultCacheKeyTemplate       = "{http.request.method} {http.request.host}{http.request.uri}"
```
Why remove the `uri.path` and `uri.query` here?

danlsgiga replied:
That's basically to use the full request URI. I was getting the same object being duplicated in the cache due to slight changes in the request URI. This fixed it for me!

danlsgiga replied:
`Use a better, more specific defaultCacheKeyTemplate to exact matches on resources that are redirected.`

danlsgiga commented:
@sillygod any chances we can get the other features I added merged?

```
- Changes how the file backend store the files by using sha256 hashes based on the entry keyWithRespectVary() content. This way, we don't waste storage space by creating duplicated temporary files every time the caddy process is restarted.
- Implemented the Length() function for the file backend, since some browsers have grief when the Content-Length is set to 0 and abort the operation causing the http request to have an empty body
```

sillygod commented:
> @sillygod any chances we can get the other features I added merged?
> 
> ```
> - Changes how the file backend store the files by using sha256 hashes based on the entry keyWithRespectVary() content. This way, we don't waste storage space by creating duplicated temporary files every time the caddy process is restarted.
> - Implemented the Length() function for the file backend, since some browsers have grief when the Content-Length is set to 0 and abort the operation causing the http request to have an empty body
> ```

Sorry for replying so late, I'll check it today.

sillygod commented on handler.go:
```diff
@@ -70,9 +70,12 @@ func (h *Handler) respond(w http.ResponseWriter, entry *Entry, cacheStatus strin
 
+	// set the response code earlier to proxy the upstream response codes
+	// This avoids the cache to respond with 200 when the upstream sent a 302 (Redirect)
+	w.WriteHeader(entry.Response.Code)
```
This part can be revert because I've fixed this with this PR #36

sillygod commented on endpoint.go:
```diff
@@ -226,14 +215,10 @@ func (c cachePurge) handlePurge(w http.ResponseWriter, r *http.Request) error {
-	conds := purgeRepl.ReplaceKnown(escapedKeyTmpl, "")
-	return c.Purge(cache, conds)
+	conds := purgeRepl.ReplaceKnown(config.CacheKeyTemplate, "")
+	return c.Purge(cache, regexp.QuoteMeta(conds))
```
You should not use the `regexp.QuoteMeaa` directly because I just want the `?` to be used as a separate of URL path and URL query not for the regular expression.

If you use `regexp.QuoteMeta here the whole conds will be escaped, that will not work as I expect.

ex.
If I want to purge all possible cache keys with a URL query with prefix `a` then the conds will look like
`GET example.com/path\?a.*`

sillygod commented on endpoint.go:
```diff
@@ -38,16 +38,14 @@ type cachePurge struct{}
 // {
 //    "method": "GET",
-//    "hots": "example.com",
+//    "host": "example.com",
```
Thanks for catching this typo
